src/System.ServiceModel.Primitives/src/Internals/System/Xml/XmlMtomWriter.cs

```diff
@@ -286,6 +286,24 @@ public override void WriteValue(IStreamProvider value)
                 Writer.WriteValue(value);
         }
 
+        public override async Task WriteValueAsync(IStreamProvider value)
+        {
+            if (value == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentNullException("value"));
+
+            if (Writer.WriteState == WriteState.Element)
+            {
+                if (_binaryDataChunks == null)
+                {
+                    _binaryDataChunks = new List<MtomBinaryData>();
+                    _contentID = GenerateUriForMimePart((_mimeParts == null) ? 1 : _mimeParts.Count + 1);
+                }
+                _binaryDataChunks.Add(new MtomBinaryData(value));
+            }
+            else
+                await Writer.WriteValueAsync(value);
+        }
```

Comment:
Making a method async causes a state machine to be generated for it and has some extra overhead. When the only async call is the last thing in a method there's an optimization you can do. You can drop the async keyword from the method and return the Task result from `Writer.WriteValueAsync(value)` to the caller. In the other code path which doesn't do anything async, return `Task.CompletedTask`.